### ISSUE TITLE
[Feature Request] Show only Important parts of URL

### ISSUE THREAD

**Issue by alexs77** [tuskyapp/Tusky#2892]
It would be very good, if Tusky would only show important parts of URLs, especially if the URL is long.

![Screenshot_20221121_194957_com keylesspalace tusky](https://user-images.githubusercontent.com/781361/203137256-3a91b7d3-1f8f-4de4-b69f-3a36ddb8603c.jpg)

This screenshot shows a toot with a very long url. It contains many parts, which aren't often important to end users. Given the limited screen real estate on mobile devices, the overly long url wastes space with no gain.

The url in question is:

https://www.stuttgarter-zeitung.de/inhalt.verkehrssuender-in-stuttgart-keine-rettungsgasse-117-fahrverbote-als-strafe.112b3762-247f-4422-9bb2-b692fabfeb42.html?utm_source=newsshowcase&utm_medium=gnews&utm_campaign=CDAqKggAIhDYhlJK7AdkzzSRhWQaqTH8KhQICiIQ2IZSSuwHZM80kYVkGqkx_DDzyoMB&utm_content=bullets​

From this URL, the scheme (https?://) should be stripped. Maybe only, if it is the normal case, nowadays (ie. https://).

Next, every query parameter (ie. starting at the `?`) should go. Especially tracking things, like `utm`.

And also the `www.` in front could go, as it's also too common.

So, the URL might be displayed like so:
stuttgarter-zeitung.de/inhalt.verkehrssuender-in-stuttgart-keine-rettungsgasse-117-fahrverbote-als-strafe.112b3762-247f-4422-9bb2-b692fabfeb42.html?…

Maybe also only display the first 20 characters of the path?

stuttgarter-zeitung.de/inhalt.verkehrssuend…

* * * *
- Tusky Version: 20.0 beta 1
- Android Version: 12
- Android Device: Huawei P20 
- Mastodon instance (if applicable): mas.to

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

**Comment by alexs77** [comment]
Another example. While it is one of my own toots, it is not staged.

On https://mas.to/@alexs77/109383798169143058, there is this.

![Screenshot_20221121_220408_com keylesspalace tusky](https://user-images.githubusercontent.com/781361/203157869-2d462102-0090-4c1e-a654-3587851a25a7.jpg)

URL: http://email.neon-free.ch/e/c/eyJlbWFpbF9pZCI6ImRnU2oxQVVEQUpfQ0E1N0NBd0dFbTBGV3p0amI5SWg3VHRvQ2VWYz0iLCJocmVmIjoiaHR0cHM6Ly93d3cubmVvbi1mcmVlLmNoL2RlL2Jsb2cvdWViZXItbmVvbi9nYXJhbnRpZXZlcmxhZW5nZXJ1bmcvP3V0bV9jYW1wYWlnbj1OTF9QQVJUTl8yMDIyMTEyMV9FUlYtV2FycmFudHktRXh0ZW5zaW9uX0RFX0VtYWlsX2ZyZWVcdTAwMjZ1dG1fY29udGVudD1OTF9QQVJUTl8yMDIyMTEyMV9FUlYtV2FycmFudHktRXh0ZW5zaW9uX0RFX0VtYWlsX2ZyZWVcdTAwMjZ1dG1fbWVkaXVtPWVtYWlsX2FjdGlvblx1MDAyNnV0bV9zb3VyY2U9Y3VzdG9tZXIuaW8iLCJpbnRlcm5hbCI6ImEzZDQwNTAxZWVkNzAxOWZjMjAzIiwibGlua19pZCI6MjE4N30/a1f5338318c86305f950fe2f83591ac2e8efb7d54e8fc22908508e25f95cc011

I have an EXTREMELY hard time believing, that anybody in the world would assume that a larger user base would really like to such a beast of URL in full — especially on a mobile phone.

**Comment by charlag** [comment]
I understand the motivation but I have concerns, because we are a mobile app and generally don't have a hover thing for the URL:

 * why should we strip specifically https and not other protocols
 * why should we strip www speicifically
 * how much of the path is relevant and how much can be used for phishing?
 * same for query parameters

Generally it feels like the server could do some of that but the same problems arise

**Comment by alexs77** [comment]
Strip "https://" and "www", as those are too common. The majority of URLs start with https://www, and so it bears no information anymore. Only display something, if it's unusual. Like http:// or ftp://. Or hostnames, that don't start with www. Refer to what browsers do in the address bar; eg. Opera:

![grafik](https://user-images.githubusercontent.com/781361/203233922-e673b191-277a-4b6a-b3ad-d8731baf141e.png)

Also consider what browsers do, when you hover over that email.neon-free.ch URL — in the status line, they don't show the full URL.

As for how much of path? If setting a limit, it has to be set somewhere. 20c seemed nice. 15c would also be good or 30c. I suggested 20c, as that seemed to be a nice compromise.

Query parameters? Well, especially utm - who cares on a mobile phone?

Also refer to what other apps/sites do. That's what users are used to. 

Server could do? No. It should sent the complete thing, so that clients have a chance of doing things.

Also think about the 2nd url I posted. Who wants to see such a url?

As for hover. A long press action could be added, which shows the full url, for those that care.

Another example: Gmail shows this for "long" URLs (like https://github.com/tuskyapp/Tusky/issues/2892#issuecomment-1322754549).
![grafik](https://user-images.githubusercontent.com/781361/203233700-cd259d33-dc2b-48cc-8364-10f21931168a.png)

Phishing: That's a non argument, I'd say. Phishing can and is often done by setting up "look alike" domains. I'd even say, that displaying the full URL makes phishing even EASIER, as users are flooded with information, that they cannot understand.

Having had a look at my spam, phishing URLs seem to look like eg. `https://hnjbvcfdre.blob.core.windows.net/hjuynbgre/jkiuyhg.html#Z29vZC5waHA%7EdD1jJmQ9MTg0OCZsPTU0NjEmYz04MTU1OSZtPTEmcD0xOTRhMTYzYTE4MmExNyZjcj03MzI%3D`. They seem to use HTML files stored on an object store somewhere. I'd say, that nobody would've been able to tell that this is a phishing URL, just by looking at it.

Or if you have a URL like https://email.newsletters.lebara.ch/e/c/eyJlbWFpbF9pZCI6ImRnVDJzZ1VEQUtIWThRR2cyUEVCQVlSWDNRQkRRTjdxWWZVWU9FQ2JpZz09IiwiaHJlZiI6Imh0dHBzOi8vd3d3LmxlYmFyYS5jaC9lbi9wcm9tbz91dG1fc291cmNlPW5ld3NsZXR0ZXJcdTAwMjZ1dG1fbWVkaXVtPWVtYWlsX2FjdGlvblx1MDAyNnV0bV9jYW1wYWlnbj1MZWJhcmFfcHJvc3BlY3Rfc3dpc3NfMTFfbm92MjAyMl9FTl9FTlx1MDAyNmNpbz04NjY1MTE2OTBmYjQ4MTdlOWIyMzY5MDA4NjNmOTllM2Q3ZGMxYjlmMDc0ZWRmOGY4YWYwODQ0OWZiNmY0NGFjX0IyMTYxIiwiaW50ZXJuYWwiOiJmNmIyMDU4MzAxOTBlOTAxYTFkOGYxMDEiLCJsaW5rX2lkIjoxODE5OH0/cd6f817f30f0af1784baa6ca66c70eae4aad3cdb8a8a58d1344b4b037331df90 — can you really tell, where there might be leaked information?

Long story short: Displaying a full URL doesn't make it more safe for the users. Rather the opposite.

**Comment by kylegoetz** [comment]
Rather than stripping, why not hide with tapping the ellipsis being an "expand" function; this is one way it's implemented in small display boxes on the web for years, overflow is turned into ellipsis that can be expanded on tap

**Comment by alexs77** [comment]
Yeah! I'd be all for that. My point is, that normally the full url
shouldn't be shown, as it wastes screen space. Your idea is even better.

That is also what Mastodon does on the web interface. Compare how this toot
***@***.***/109425421918399537 is displayed on mas.to
on a phone and on tusky. Maybe tusky should copy the logic. This way, it
would look identical. Users wouldn't have to readjust.

**Comment by kylegoetz** [comment]
Which do we think is better after shortening a URL longer than some hard-coded threshold:

1. click URL, visit URL; click ellipsis, expand URL (two buttons could be confusing)
2. click URL or ellipsis, expand URL; click expanded URL, visits URL (two clicks to open URL, could be annoying)

**Comment by alexs77** [comment]
How about mimicking what Mastodon or the browser does?

On the attached screenshots, I long pressed on a shortened URL. My browser
(Firefox on Android in my case) then shows the long url in the title of the
popup.

So do Opera Mini and Dolphin (screenshot with green popup).

Because of that, I'd go with something like your 1st suggestion. My hope
is, that users would get that, because that's what they're used to anyways
already.

But it's of course also true, that not so techie folks might not know that.
But would a not-techie-person be able to read a long url and spot phishing
in there? I highly doubt that.

**Comment by alexs77** [comment]
My coding fu really isn't good enough, but would it be possible to fix this?

Another example.

![grafik](https://user-images.githubusercontent.com/781361/212896817-a623a672-dbca-4fc6-9e57-1b559b11db16.png)

Due to the ultra long URLs, not all of the text is actually readable.

URL: https://mas.to/@alexs77/109704473317009700

**Comment by scollovati** [comment]
The Fedilab app actually has a feature flag (in the privacy section) where a user can choose to strip all UTM query parameters.

 **This is a good thing both from UI and a privacy perspectives.**

**Comment by alexs77** [comment]
Is there actually any progress in making Tusky better usable, as far as that is concerned? There's no security advantage in wasting screen space and hiding "normal" text.

Reg. stripping UTM: Might be nice, but sometimes it's also good to leave UTM intact. And it also won't solve issues as shown with me "neon" link in <https://github.com/tuskyapp/Tusky/issues/2892#issuecomment-1322641062>.
